dm/syncer/compactor.go

```diff
@@ -61,48 +66,62 @@ func compactorWrap(inCh chan *job, syncer *Syncer) chan *job {
 
 // run runs a compactor instance.
 func (c *compactor) run() {
-	for j := range c.inCh {
-		metrics.QueueSizeGauge.WithLabelValues(c.task, "compactor_input", c.source).Set(float64(len(c.inCh)))
-
-		if j.tp == flush {
-			c.flushBuffer()
-			c.outCh <- j
-			continue
-		}
-
-		// set safeMode when receive first job
-		if len(c.buffer) == 0 {
-			c.safeMode = j.dml.safeMode
-		}
-		// if dml has no PK/NOT NULL UK, do not compact it.
-		if j.dml.identifyColumns() == nil {
-			c.buffer = append(c.buffer, j)
-			continue
-		}
-
-		// if update job update its identify keys, turn it into delete + insert
-		if j.dml.op == update && j.dml.updateIdentify() {
-			delDML, insertDML := updateToDelAndInsert(j.dml)
-			delJob := j.clone()
-			delJob.tp = del
-			delJob.dml = delDML
-
-			insertJob := j.clone()
-			insertJob.tp = insert
-			insertJob.dml = insertDML
-
-			c.compactJob(delJob)
-			c.compactJob(insertJob)
-		} else {
-			c.compactJob(j)
-		}
-
-		failpoint.Inject("SkipFlushCompactor", func() {
-			failpoint.Continue()
-		})
-		// if no inner jobs, buffer is full or outer channel empty, flush the buffer
-		if len(c.inCh) == 0 || len(c.buffer) >= c.bufferSize || len(c.outCh) == 0 {
-			c.flushBuffer()
+	for {
+		select {
+		case j, ok := <-c.inCh:
+			if !ok {
+				return
+			}
+			metrics.QueueSizeGauge.WithLabelValues(c.task, "compactor_input", c.source).Set(float64(len(c.inCh)))
+
+			if j.tp == flush {
+				c.flushBuffer()
+				c.outCh <- j
+				continue
+			}
+
+			// set safeMode when receive first job
+			if len(c.buffer) == 0 {
+				c.safeMode = j.dml.safeMode
+			}
+			// if dml has no PK/NOT NULL UK, do not compact it.
+			if j.dml.identifyColumns() == nil {
+				c.buffer = append(c.buffer, j)
+				continue
+			}
+
+			// if update job update its identify keys, turn it into delete + insert
+			if j.dml.op == update && j.dml.updateIdentify() {
+				delDML, insertDML := updateToDelAndInsert(j.dml)
+				delJob := j.clone()
+				delJob.tp = del
+				delJob.dml = delDML
+
+				insertJob := j.clone()
+				insertJob.tp = insert
+				insertJob.dml = insertDML
+
+				c.compactJob(delJob)
+				c.compactJob(insertJob)
+			} else {
+				c.compactJob(j)
+			}
+
+			failpoint.Inject("SkipFlushCompactor", func() {
+				failpoint.Continue()
+			})
+			// if the number of outer jobs is zero or buffer is full, flush the buffer
+			if len(c.outCh) == 0 || len(c.buffer) >= c.bufferSize {
+				c.flushBuffer()
+			}
+			// if no inner jobs and the number of outer jobs is zero, flush the buffer
+		case <-time.After(waitTime):
+			failpoint.Inject("SkipFlushCompactor", func() {
+				failpoint.Continue()
+			})
+			if len(c.outCh) == 0 {
```

Comment:
I think when there are no new job in waitTime, we should always do flush here.

Reply:
The reason for changing this behavior is that if the downstream is very slow(the outer channel has jobs), but the job sent by upstream is not fast(maybe slow than 10ms?). At this time, if we still flush jobs, the compactor buffer will always be empty, which does not conform to the logic of compact jobs when the downstream is slow.

Reply:
But if there is no job in the upstream 10ms, seems it doesn't matter if it is not compact?

Reply:
agree with @GMHDBJD 
just a question, when downstream is slow, we keep still flush when buffer is full and this will increase the pressure of downsteram, maybe we can incr buffer size here🤔

Reply:
May cause OOM, buffer size too big will also increase the time for stop-task.

Reply:
07a89a7